quantities/parser_body.hpp

```diff
@@ -12,125 +13,194 @@ namespace principia {
 namespace quantities {
 namespace internal_parser {
 
-// The patterns that we parse here have the form:
-//    U^n/V^m
-//  when U and V are (final) unit names and n and m are integers.
+using si::AstronomicalUnit;
+using si::Day;
+using si::Degree;
+using si::Kilo;
+using si::Metre;
+using si::Radian;
+using si::Second;
 
-template<typename T>
-using ParseUnitFunction = T(*)(std::string const& s);
+using RuntimeDimensions = std::array<std::int64_t, 8>;
 
-template<typename T, int exponent>
-Exponentiation<T, exponent> ParseExponentiationUnit(std::string const& s) {
-  int const first_carret = s.find('^');
-  int const last_nonblank = s.find_last_not_of(' ', first_carret - 1);
-  CHECK_NE(std::string::npos, last_nonblank);
+template<typename Q>
+struct ExtractDimensions {};
 
-  char* interpreted_end;
-  const char* interpreted_begin = s.c_str() + first_carret + 1;
-  int const actual_exponent = std::strtol(interpreted_begin,
-                                          &interpreted_end,
-                                          10);
-  int const interpreted = interpreted_end - interpreted_begin;
-  CHECK_LT(0, interpreted) << "invalid exponent " << s;
-  CHECK_EQ(exponent, actual_exponent);
-
-  return Pow<exponent>(ParseUnit<T>(s.substr(0, last_nonblank + 1)));
+template<>
+struct ExtractDimensions<double> {
+  constexpr static RuntimeDimensions dimensions = {0, 0, 0, 0, 0, 0, 0, 0};
+};
+
+template<std::int64_t LengthExponent,
+         std::int64_t MassExponent,
+         std::int64_t TimeExponent,
+         std::int64_t CurrentExponent,
+         std::int64_t TemperatureExponent,
+         std::int64_t AmountExponent,
+         std::int64_t LuminousIntensityExponent,
+         std::int64_t AngleExponent>
+struct ExtractDimensions<
+    Quantity<internal_quantities::Dimensions<LengthExponent,
+                                             MassExponent,
+                                             TimeExponent,
+                                             CurrentExponent,
+                                             TemperatureExponent,
+                                             AmountExponent,
+                                             LuminousIntensityExponent,
+                                             AngleExponent>>> {
+  constexpr static RuntimeDimensions dimensions = {LengthExponent,
+                                                   MassExponent,
+                                                   TimeExponent,
+                                                   CurrentExponent,
+                                                   TemperatureExponent,
+                                                   AmountExponent,
+                                                   LuminousIntensityExponent,
+                                                   AngleExponent};
+};
+
+struct Unit {
```

Comment:
Add a templated factory that constructs this from a `Quantity` so we can write `ExtractUnit(AstronomicalUnit)` rather than `{ExtractDimensions<Length>::dimensions, AstronomicalUnit / Metre}` which is redundant and error-prone.